### ISSUE TITLE
rustc::lint could use some helper functions for working with macros

### ISSUE THREAD

**Issue by llogiq** [rust-lang/rust#33177]
(See #22451 and possibly others)

Currently we have a few functions in [clippy](https://github.com/Manishearth/rust-clippy)`/src/util/mod.rs` to deal with macros. Namely we can check if some span stems from a macro expansion (though that one could need some work), or if some span was expanded by a given macro (by name) somewhere or directly, or if two spans are in the same macro expansion. This is often useful for readability lints which should not be invoked within macro-expanded code (unless perhaps the code before expansion had the same issue, but that is as of yet hard to detect).

I think those functions belong in the rustc::lint crate. This would both benefit internal lints and all other lints outside of clippy which could then make use of those helpers.

However, clippy is currently licensed under MPL, so we'd need agreement from the authors (of the respective code lines) to move this code into rustc proper. Also we'll want to look first if those functions are ready to be moved and clean up any technical debt we may have incurred so far.

cc @Manishearth

**Comment by Manishearth** [comment]
As far as licensing is concerned, feel free to relicense my contributions to clippy as Apache v2, MIT, or dual Apache v2/MIT.


... should have put more thought into the license when picking it initially. It has made it easier to copy code into Servo (which I do every now and then, and Servo is MPL), but I'm not sure what the situation with copying to Rust without needing to add a license is. I'm okay with relicensing the whole of clippy, too.

**Comment by Manishearth** [comment]
I actually looked through Rust's lints and most of them don't _need_ macro checks; they all make sense from within a macro or across macro boundaries. Some of them like the unused_ ones just need your macro to have underscores or allow annotations. I don't think there's a pressing need for macro checks for any of them, however I would like to move macro checks into rustc::lint so that other lint writers can use them.

**Comment by steveklabnik** [comment]
Triage: A *lot* has changed since 2016. Do either of you feel this is still an issue?

**Comment by Manishearth** [comment]
Yes, I think there are still lints that need macro checks. I recall coming across this relatively recently.

**Comment by workingjubilee** [comment]
"How to identify you are linting into a macro and stop linting if you are an `unused_whatever` style of lint" is not discussed in the [rustc dev guide section on diagnostics](https://rustc-dev-guide.rust-lang.org/diagnostics.html).
